### PR TITLE
Updates in Trainer to support partial checkpointing for SM Model Parallel library 

### PR DESCRIPTION
**Pull request by cavdard**
# What does this PR do?

- Adds 2 new training args(`smp_save_partial` and `smp_load_partial`) to support partial checkpointing with SMP
- Uses the right ranks for partial checkpoint saving in should_save.
- Uses `local_state_dict()` with partial checkpoint saving.
- Uses `smp.save` instead of `torch.save` when partial checkpoint saving is enabled. 
- Uses `smp.load` instead of `torch.load` when partial checkpoint loading is enabled.  Reorders partial checkpoint loading to happen after wrapping of model, since `smp.load` can only load to a smp model.
- Updated checks for the existence of checkpoint files since smp partial checkpoints contain postfixes in addition to filename(example: filename_0_0 or filename_0_0_0).
- `smp_gather` is causing increased memory usage on GPU0 when tensor parallelism is enabled. Switches to `distributed_concat` for ddp. 
- adds `load_best_model_at_end` support for SMP.

Fixes # (issue)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -1361,32 +1361,39 @@ def save_pretrained(
...
-        # Clean the folder from a previous save
-        for filename in os.listdir(save_directory):
-            full_filename = os.path.join(save_directory, filename)
-            if filename.startswith(WEIGHTS_NAME[:-4]) and os.path.isfile(full_filename):
-                os.remove(full_filename)
+            # Clean the folder from a previous save
+            for filename in os.listdir(save_directory):
+                full_filename = os.path.join(save_directory, filename)
+                if filename.startswith(WEIGHTS_NAME[:-4]) and os.path.isfile(full_filename):
+                    os.remove(full_filename)
```
Comment (rahul003) [inline src/transformers/modeling_utils.py:1378]: Is this not needed for SMP?

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1715,16 +1745,22 @@ def _save_checkpoint(self, model, trial, metrics=None):
...
             if smp.rdp_rank() == 0:
-                # Consolidate the state dict on all processed of rdp_rank 0
-                opt_state_dict = self.optimizer.state_dict()
-                # Save it and the scheduler on the main process
-                if self.args.should_save:
+            # Consolidate the state dict on all processed of rdp_rank 0
+                if self.args.smp_save_partial:
+                    opt_state_dict = self.optimizer.local_state_dict()
+                else:
+                    opt_state_dict = self.optimizer.state_dict()
```
Comment (rahul003) [inline src/transformers/trainer.py:1752]: We wanted to standardize on gather_if_shard=False here. 

When we do that, which processes need to save the partial state dict changes.

if shard_optimizer_state, then all processes
else rdp_rank==0

**Comment by LysandreJik** [comment]
Hey! It seems a bad rebase/merge happened on your PR. Usually, closing this PR and opening a new one from the same branch solves the problem.

**Comment by cavdard** [comment]
Closing this PR. 
Created a new [PR](https://github.com/huggingface/transformers/pull/16734)

**Pull request closed without merging**